### PR TITLE
clean API for moveAt, fix indexes/lengths to be type of size_t, fix length declarations

### PR DESCRIPTION
**Pull request by 9il**

### CODE REVIEW

**Comment by greenify** [comment]
On a different matter - if we expose `swapAt` in std.algorithm, shouldn't we expose `moveAt` there too?

https://dlang.org/phobos/std_algorithm_mutation.html
http://dlang.org/phobos/std_range_primitives.html

**Comment by 9il** [comment]
... and other `move` primitives. This will break API :-(

**Comment by greenify** [comment]
> ... and other move primitives. This will break API :-(

But if we agree that this would make sense, can't we import them as alias and add a deprecated attribute warning?

std.algorithm.mutation has the following description:

> This is a submodule of std.algorithm. It contains generic mutation algorithms.

Whereas std.range.primitives states

> Finally, it includes some convenience functions for manipulating ranges:

> moveFront	Removes the front element of a range.
> moveBack	Removes the back element of a bidirectional range.
> moveAt	        Removes the i'th element of a random-access range.

Apart from that we might consider these two templates:

> hasMobileElements	Tests if a given range's elements can be moved around using the primitives moveFront, moveBack, or moveAt.

> hasSwappableElements	Tests if a range is a forward range with swappable elements.

... it even imports swap from std.algorithm

**Comment by 9il** [comment]
> But if we agree that this would make sense, can't we import them as alias and add a deprecated attribute warning?

Does this would break code if both modules are imported?

**Comment by 9il** [comment]
Yes: from std.algorithm
```
// Explicitly undocumented. It will be removed in March 2017. @@@DEPRECATED_2017-03@@@
deprecated("Please use std.functional.forward instead.")
alias forward = std.functional.forward;
```

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by greenify** [comment]
Seems like just changing the API isn't that easy:

```
std/range/package.d(291): Error: template std.range.primitives.moveAt cannot deduce function from argument types !()(Result, ulong), candidates are:
std/range/primitives.d(1946):        std.range.primitives.moveAt(R)(R r, size_t i)
std/range/package.d(313): Error: template instance std.range.retro!(Result).retro.Result!() error instantiating
std/algorithm/iteration.d(3261):        instantiated from here: retro!(Result)
std/range/package.d(291): Error: template std.range.primitives.moveAt cannot deduce function from argument types !()(Result!(), ulong), candidates are:
std/range/primitives.d(1946):        std.range.primitives.moveAt(R)(R r, size_t i)
std/range/package.d(313): Error: template instance std.range.retro!(Result!()).retro.Result!() error instantiating
std/algorithm/iteration.d(3051):        instantiated from here: retro!(Result!())
std/algorithm/iteration.d(3262):        instantiated from here: splitter!("a == b", Result!(), long)
std/algorithm/iteration.d(3709): Error: template instance std.algorithm.iteration.splitter!("a == b", Result, long[]) error instantiating
```

**Comment by 9il** [comment]
@DmitryOlshansky, I have changed indexes/length to be always `size_t`. This will reduce template bloat and makes Phobos more predictable

**Comment by DmitryOlshansky** [comment]
Nice!

File: std/range/package.d [hunk 1]
```diff
@@ -213,7 +213,7 @@ if (isBidirectionalRange!(Unqual!Range))
     }
     else
     {
-        static struct Result()
+        static struct Result
```
Comment (DmitryOlshansky) [inline std/range/package.d]: Please keep template, it may have conseqence on inlining

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled off

File: std/range/package.d (hunk 1 above)
Comment (9il) [inline std/range/package.d]: OK, where I can find an issue for this bug?

Comment (DmitryOlshansky) [inline std/range/package.d]: No idea, I've been using 0-arg templates for "please inline me" with great success in std.regex

**Comment by 9il** [comment]
Fixed and rebased

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by schveiguy** [comment]
Should there be a changelog entry for these things? I think especially the change to size_t for iota.length should be noted as it could break some code.

**Comment by 9il** [comment]
Is `pending changelog` is correct label for that?

**Comment by schveiguy** [comment]
Hm... I'd say yes, but this has already been merged.

**Comment by 9il** [comment]
I will add change log for all my PRs (this would be big change log)

**Comment by schveiguy** [comment]
You only need to add an entry if there is no bug closed. The closed bugs are used to automatically generate entries.

**Comment by wilzbach** [comment]
> Should there be a changelog entry for these things? I think especially the change to size_t for iota.length should be noted as it could break some code.

We should probably document that `swapAt` is now public too - should I quickly submit a PR?

**Comment by 9il** [comment]
> We should probably document that swapAt is now public too - should I quickly submit a PR?

Yes

**Comment by wilzbach** [comment]
See #4220 ;-)

**Pull request merged**